plutus-tx-plugin/src/PlutusTx/Compiler/Expr.hs

```diff
@@ -710,27 +715,31 @@ compileExpr e = withContextM 2 (sdToTxt $ "Compiling expr:" GHC.<+> GHC.ppr e) $
                     Nothing      -> throwSd UnsupportedError $ "Cannot case on a value of type:" GHC.<+> GHC.ppr scrutineeType
                 dcs <- getDataCons tc
 
+                match <- getMatchInstantiated scrutineeType
+                let matched = PIR.Apply () match scrutinee'
+
                 -- See Note [Case expressions and laziness]
-                isPureAlt <- forM dcs $ \dc ->
-                    let (_, vars, body) = findAlt dc alts t
-                    in if null vars then PIR.isPure (const PIR.NonStrict) <$> compileExpr body else pure True
+                alternatives <- forM dcs $ \dc -> do
+                    let alt = findAlt dc alts t
+                        -- these are the instantiated type arguments, e.g. for the data constructor Just when
+                        -- matching on Maybe Int it is [Int] (crucially, not [a])
+                        instArgTys = GHC.dataConInstOrigArgTys dc argTys
+                    (nonDelayedAlt, delayedAlt) <- compileAlt alt instArgTys
+                    return (alt, nonDelayedAlt, delayedAlt)
+                isPureAlt <- forM alternatives $ \((_, vars, _), nonDelayed, _) ->
+                    pure $ if null vars
+                        then PIR.isPure (const PIR.NonStrict) nonDelayed
```

Comment:
you can just use this branch unconditionally now. Also then you don't need to return `alt` on line 728 either.